subsys/usb/device_next/class/usbd_hid_api.c

```diff
@@ -0,0 +1,220 @@
+{
+	const struct hid_ops *legacy_ops = get_legacy_ops(dev);
+	struct usb_setup_packet setup = {
+		.bmRequestType = 0,
+		.bRequest = 0,
+		.wValue = (type << 8) | id,
+		.wIndex = 0,
+		.wLength = len,
+	};
+	uint8_t *d = buf;
+	int l = len;
+
+	if (legacy_ops != NULL && legacy_ops->get_report != NULL) {
+		return legacy_ops->get_report(dev, &setup, &l, &d);
+	}
+
+	return -ENOTSUP;
+}
+
+int wrapper_set_report(const struct device *dev,
+		       const uint8_t type, const uint8_t id,
+		       const uint16_t len, const uint8_t *const buf)
+{
+	const struct hid_ops *legacy_ops = get_legacy_ops(dev);
+	struct usb_setup_packet setup = {
+		.bmRequestType = 0,
+		.bRequest = 0,
+		.wValue = (type << 8) | id,
+		.wIndex = 0,
+		.wLength = len,
+	};
+	uint8_t *d = (void *)buf;
+	int l = len;
+
+	if (legacy_ops != NULL && legacy_ops->set_report != NULL) {
+		return legacy_ops->get_report(dev, &setup, &l, &d);
+	}
+
+	return -ENOTSUP;
+}
+
+void wrapper_set_idle(const struct device *dev,
+		      const uint8_t id, const uint32_t duration)
+{
+	if (id != 0U) {
+		LOG_ERR("Set Idle for %p ID %u duration %u cannot be propagated",
+			dev, id, duration);
+	}
+}
+
+void wrapper_set_protocol(const struct device *dev, const uint8_t proto)
+{
+	const struct hid_ops *legacy_ops = get_legacy_ops(dev);
+
+	if (legacy_ops != NULL && legacy_ops->protocol_change != NULL) {
+		legacy_ops->protocol_change(dev, proto);
+	}
+}
+
+void wrapper_input_report_done(const struct device *dev)
+{
+	const struct hid_ops *legacy_ops = get_legacy_ops(dev);
+
+	if (legacy_ops == NULL) {
+		return;
+	}
+
+	if (legacy_ops->int_in_ready) {
+		LOG_ERR("Input report done for %p", dev);
+		legacy_ops->int_in_ready(dev);
+	}
+}
+
+void wrapper_output_report(const struct device *dev,
+			   const uint16_t len, const uint8_t *const buf)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(len);
+	ARG_UNUSED(buf);
+
+	__ASSERT(false, "Output report callback is not supported");
+}
+
+static struct hid_device_ops wrapper_ops = {
+	.get_report = wrapper_get_report,
+	.set_report = wrapper_set_report,
+	.set_idle = wrapper_set_idle,
+	.set_protocol = wrapper_set_protocol,
+	.input_report_done = wrapper_input_report_done,
+	.output_report = wrapper_output_report,
+};
+
+int hid_int_ep_write(const struct device *dev,
+		     const uint8_t *data, uint32_t data_len, uint32_t *bytes_ret)
+{
+	int ret;
+
+	ret = hid_device_submit_report(dev, data_len, data, false);
+	if (bytes_ret != NULL) {
+		*bytes_ret = ret == 0 ? data_len : 0;
+	}
+
+	return ret;
+}
+
+int hid_int_ep_read(const struct device *dev,
+		    uint8_t *data, uint32_t max_data_len, uint32_t *ret_bytes)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(data);
+	ARG_UNUSED(max_data_len);
+	ARG_UNUSED(ret_bytes);
+
+	LOG_ERR("Not supported");
+
+	return -ENOTSUP;
+}
+
+int usb_hid_set_proto_code(const struct device *dev, uint8_t proto_code)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(proto_code);
+
+	LOG_DBG("Protocol code is set using DT property boot-protocol-code");
```

Comment:
But it is not an error, it just there to get it compiled.

Reply:
I mean that an application can no longer rely on this function after it switches to USB next stack (it actually does nothing here). It would be great if we could let user know about it. Debug log would not be displayed in most of the default app configurations I think (it may also be easily overlooked). Maybe we could e.g. change it to: `LOG_WRN_ONCE` ?

Reply:
This it not called in a loop, so why should we use LOG_WRN_ONCE? I changed it to LOG_WRN().

Reply:
You are right that there should not be too many USB HID class instances. `LOG_WRN` sounds good